crawler/src/main/java/de/unidisk/services/ProjectGenerationService.java

```diff
@@ -0,0 +1,69 @@
+package de.unidisk.services;
+
+import de.unidisk.contracts.exceptions.DuplicateException;
+import de.unidisk.entities.hibernate.Project;
+import de.unidisk.contracts.exceptions.DuplicateException;
+import de.unidisk.contracts.repositories.IKeywordRepository;
+import de.unidisk.contracts.repositories.ITopicRepository;
+import de.unidisk.contracts.repositories.params.keyword.CreateKeywordParams;
+import de.unidisk.contracts.repositories.params.project.CreateProjectParams;
+import de.unidisk.contracts.services.IProjectService;
+import de.unidisk.contracts.services.recommendation.IKeywordRecommendationService;
+import de.unidisk.contracts.services.recommendation.KeywordRecommendation;
+import de.unidisk.contracts.services.recommendation.RecommendationResult;
+import de.unidisk.entities.hibernate.Keyword;
+import de.unidisk.entities.hibernate.Project;
+import de.unidisk.contracts.repositories.IProjectRepository;
+import de.unidisk.entities.hibernate.ProjectSubtype;
+import de.unidisk.entities.hibernate.Topic;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class ProjectGenerationService {
+    IProjectRepository projectRepository;
+    ITopicRepository topicRepository;
+    IKeywordRepository keywordRepository;
+    IKeywordRecommendationService keywordRecommendationService;
+
+
+    public ProjectGenerationService(IProjectRepository projectRepository, ITopicRepository topicRepository, IKeywordRepository keywordRepository, IKeywordRecommendationService keywordRecommendationService) {
+        this.projectRepository = projectRepository;
+        this.topicRepository = topicRepository;
+        this.keywordRepository = keywordRepository;
+        this.keywordRecommendationService = keywordRecommendationService;
+    }
+
+    public Project generateProjectFromTopics(String projectId) throws DuplicateException {
+
+        final Optional<Project> optionalProject = projectRepository.getProjectDetails(projectId);
+        if(!optionalProject.isPresent())
+                return null;
```

Comment:
Bleib mal bei der Optionalrückgabe. Um das Bewusstsein zu schaffen, dass Project hier null werden kann, sind ja Optionals da.

Reply:
Ich glaube, es wäre sinnig mit einem ifPresent(project -> meaningfulFunctionName(project)) zu arbeiten. Dann musst du zwar die Hauptaufgabe nochmal in eine Funktion auslagern, aber musst nicht erst mit "isPresent()" und "get()" das Project aus dem Optional holen

Reply:
Werfe jetzt auch hier exceptions